### PR TITLE
[update] 添加 spark streaming listener

### PR DESCRIPTION
**Pull request by liuweixiao**

### CODE REVIEW

File: streamingpro-mlsql/src/main/java/streaming/core/LoalSparkServiceApp.scala [hunk 1]
```diff
@@ -23,8 +23,9 @@ object LocalSparkServiceApp {
       "-spark.sql.codegen.wholeStage", "true",
       "-spark.kryoserializer.buffer.max", "2000m",
       "-streaming.udf.clzznames","streaming.crawler.udf.Functions",
-      "-streaming.driver.port", "9003"
-
+      "-streaming.driver.port", "9003",
+      "streaming.graphiteHost", "localhost",
```
Comment (cfmcgrady) [inline streamingpro-mlsql/src/main/java/streaming/core/LoalSparkServiceApp.scala:27]: 参数格式不对，另外这个参数不要添加到这里，一般测试机没有graphite服务，写到文档说明里。

File: streamingpro-mlsql/pom.xml [hunk 2]
```diff
@@ -340,6 +340,11 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>com.dxy.data</groupId>
+            <artifactId>module-spark-monitor</artifactId>
+            <version>1.0-SNAPSHOT</version>
+        </dependency>
```
Comment (cfmcgrady) [inline streamingpro-mlsql/pom.xml:347]: 不要依赖内部组件

File: streamingpro-mlsql/src/main/java/streaming/core/strategy/platform/SparkStructuredStreamingRuntime.scala [hunk 3]
```diff
@@ -55,7 +57,33 @@ class SparkStructuredStreamingRuntime(_params: JMap[Any, Any]) extends Streaming
...
+    spark.conf.set("spark.sql.streaming.metricsEnabled", "true")
+    if (params.containsKey("streaming.graphiteHost") && params.containsKey("streaming.graphitePort")) {
+      val graphitePreix = "spark.StructuredStreaming"
+      val graphiteHost = params.get("streaming.graphiteHost").toString
+      val graphitePort = params.get("streaming.graphitePort").toString.toInt
+
+
+      spark.streams.addListener(new StreamingQueryListener() {
+        override def onQueryStarted(queryStarted: QueryStartedEvent): Unit = {
+          logger.error(s"""spark application ${spark.sparkContext.appName} query start, its runid is ${queryStarted.name} """)
```
Comment (cfmcgrady) [inline streamingpro-mlsql/src/main/java/streaming/core/strategy/platform/SparkStructuredStreamingRuntime.scala:71]: error?

File: streamingpro-mlsql/src/main/java/streaming/core/strategy/platform/SparkStructuredStreamingRuntime.scala [hunk 4]
```diff
@@ -55,7 +57,33 @@ class SparkStructuredStreamingRuntime(_params: JMap[Any, Any]) extends Streaming
...
+        }
+        override def onQueryTerminated(queryTerminated: QueryTerminatedEvent): Unit = {
+          logger.error(s"""spark application ${spark.sparkContext.appName} query down, its runid is ${queryTerminated.id} """)
```
Comment (cfmcgrady) [inline streamingpro-mlsql/src/main/java/streaming/core/strategy/platform/SparkStructuredStreamingRuntime.scala:74]: error?

**Comment by allwefantasy** [comment]
SparkStructuredStreamingRuntime is used as Application mode, that means it runs accrording to a JSon configuration file. But note that application mode is a lack of maintaining in StreamingPro.

For now ,SparkRuntime is the only runtime to run as Server mode. 

We strongly recommend you to  contribute the target feature with two steps:

1. Create a mechanism of  component registration so  you can be easy to register user-defined spark listeners. Say that we want to  register graphite output , with this new mechanism,  you can do like following:

```
-streaming.ss.metrics.impl=streaming.ss.metrics.graphite.OutputImpl
-streaming.ss.metrics.graphite.address=[host:port]
```
2. Create a package named "streaming.ss.metrics.graphite" where you can place the real implementation how to write spark structured streaming message to graphite.

**Comment by wuheyi** [comment]
-streaming.ss.metrics.impl 可以指定多个sink，用逗号间隔
sink可以使用缩写，实现可以参考Runtime
image

**Pull request closed without merging**